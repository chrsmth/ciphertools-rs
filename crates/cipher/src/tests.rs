use once_cell::sync::Lazy;

#[macro_export]
macro_rules! cipher_test {
  ($context:expr, $plaintext:expr, $encipher_key:expr, $name:ident) => {
    #[test]
    #[allow(non_snake_case)]
    fn $name() {
      use insta::assert_snapshot;
      let ciphertext = $context.encipher(&$plaintext, &$encipher_key);
      let plaintext = $context.decipher(&ciphertext, &$encipher_key);

      assert_eq!(plaintext, $plaintext);
      assert_snapshot!(ciphertext);
    }
  };
}

macro_rules! lazy_read_to_string {
  ($path:literal) => {
    Lazy::new(|| {
      std::fs::read_to_string($path)
        .unwrap_or_else(|e| panic!("Failed to read {}: {}", $path, e))
    })
  };
}

pub static LATIN_LOWERCASE: &[char; 26] = &[
  'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o',
  'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
];
pub static RUSSIAN_LOWERCASE: &[char; 33] = &[
  'а', 'б', 'в', 'г', 'д', 'е', 'ё', 'ж', 'з', 'и', 'й', 'к', 'л', 'м', 'н',
  'о', 'п', 'р', 'с', 'т', 'у', 'ф', 'х', 'ц', 'ч', 'ш', 'щ', 'ъ', 'ы', 'ь',
  'э', 'ю', 'я',
];
pub static GREEK_LOWERCASE: &[char; 24] = &[
  'α', 'β', 'γ', 'δ', 'ε', 'ζ', 'η', 'θ', 'ι', 'κ', 'λ', 'μ', 'ν', 'ξ', 'ο',
  'π', 'ρ', 'σ', 'τ', 'υ', 'φ', 'χ', 'ψ', 'ω',
];
pub static DIGITS: &[char; 10] =
  &['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
pub static SYMBOLS: &[char; 32] = &[
  '!', '@', '#', '$', '%', '^', '&', '*', '(', ')', '_', '-', '+', '=', '{',
  '}', '[', ']', ':', ';', '<', '>', '?', ',', '.', '/', '\\', '|', '`', '~',
  '\"', '\'',
];

pub static QUICK_BROWN_FOX: Lazy<String> =
  lazy_read_to_string!("./test-texts/plaintexts/the_quick_brown_fox.txt");

pub static QUICKBROWNFOX: Lazy<String> =
  lazy_read_to_string!("./test-texts/plaintexts/thequickbrownfox.txt");
