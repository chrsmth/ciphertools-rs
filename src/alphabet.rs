use ahash::AHashMap;

static LATIN: &[char; 26] = &[
  'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o',
  'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
];

static RUSSIAN: &[char; 33] = &[
  'а', 'б', 'в', 'г', 'д', 'е', 'ё', 'ж', 'з', 'и', 'й', 'к', 'л', 'м', 'н',
  'о', 'п', 'р', 'с', 'т', 'у', 'ф', 'х', 'ц', 'ч', 'ш', 'щ', 'ъ', 'ы', 'ь',
  'э', 'ю', 'я',
];

static GREEK: &[char; 24] = &[
  'α', 'β', 'γ', 'δ', 'ε', 'ζ', 'η', 'θ', 'ι', 'κ', 'λ', 'μ', 'ν', 'ξ', 'ο',
  'π', 'ρ', 'σ', 'τ', 'υ', 'φ', 'χ', 'ψ', 'ω',
];

#[derive(Debug, Clone)]
pub struct Alphabet {
  chars: Vec<char>,
  indexes: AHashMap<char, usize>,
}

impl Alphabet {
  pub fn russian() -> Alphabet {
    Alphabet::from_iter(RUSSIAN.iter().cloned())
  }

  pub fn greek() -> Alphabet {
    Alphabet::from_iter(GREEK.iter().cloned())
  }

  pub fn latin() -> Alphabet {
    Alphabet::from_iter(LATIN.iter().cloned())
  }

  pub fn iter(&self) -> impl Iterator<Item = char> + use<'_> {
    self.chars.iter().copied()
  }
}

impl Default for Alphabet {
  fn default() -> Self {
    Alphabet::latin()
  }
}

impl FromIterator<char> for Alphabet {
  fn from_iter<I: IntoIterator<Item = char>>(iter: I) -> Alphabet {
    let chars: Vec<_> = iter.into_iter().collect();
    let indexes = AHashMap::from_iter(
      chars.iter().cloned().enumerate().map(|(i, c)| (c, i)),
    );

    Alphabet { chars, indexes }
  }
}

impl Alphabet {
  pub fn get(&self, i: usize) -> Option<char> {
    self.chars.get(i).map(|x| x.to_owned())
  }

  pub fn get_index(&self, c: char) -> Option<usize> {
    self.indexes.get(&c).map(|x| x.to_owned())
  }

  pub fn len(&self) -> usize {
    self.chars.len()
  }

  pub fn _contains(&self, c: char) -> bool {
    self.indexes.contains_key(&c)
  }

  pub fn add(&self, a: char, b: char) -> char {
    self
      .get_index(a)
      .zip(self.get_index(b))
      .and_then(|(ai, bi)| self.get((ai + bi) % self.len()))
      .unwrap_or(a)
  }

  pub fn sub(&self, a: char, b: char) -> char {
    self
      .get_index(a)
      .zip(self.get_index(b))
      .and_then(|(ai, bi)| self.get((self.len() + ai - bi) % self.len()))
      .unwrap_or(a)
  }
}
