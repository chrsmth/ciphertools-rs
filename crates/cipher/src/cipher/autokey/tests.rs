use super::*;
use crate::alphabet::Alphabet;
use crate::cipher_test;
use crate::tests::*;
use once_cell::sync::Lazy;

pub static CONTEXT_INCLUDE_WHITESPACE: Lazy<Autokey> =
  Lazy::new(|| Autokey::new(Alphabet::latin(), false));
pub static CONTEXT_SKIP_WHITESPACE: Lazy<Autokey> =
  Lazy::new(|| Autokey::new(Alphabet::latin(), true));

pub static KEY_A: Lazy<AutokeyKey> =
  Lazy::new(|| AutokeyKey::new("key".to_string()));

#[test]
fn key_try_new() {
  assert!(AutokeyKey::try_new(
    "hello".to_string(),
    &CONTEXT_INCLUDE_WHITESPACE
  )
  .is_ok());
  assert!(matches!(
    AutokeyKey::try_new("gr8ness".to_string(), &CONTEXT_INCLUDE_WHITESPACE),
    Err(ParseError::InvalidChar('8')),
  ));
}

cipher_test!(
  CONTEXT_INCLUDE_WHITESPACE,
  *QUICK_BROWN_FOX,
  KEY_A,
  include_whitespace
);

cipher_test!(
  CONTEXT_SKIP_WHITESPACE,
  *QUICK_BROWN_FOX,
  KEY_A,
  skip_whitespace
);
