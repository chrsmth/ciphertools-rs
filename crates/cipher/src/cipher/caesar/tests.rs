use super::*;
use crate::alphabet::Alphabet;
use crate::cipher_test;
use crate::tests::*;
use once_cell::sync::Lazy;

pub static CONTEXT: Lazy<Caesar> = Lazy::new(|| Caesar::new(Alphabet::latin()));

pub static KEY_A: Lazy<CaesarKey> = Lazy::new(|| CaesarKey::new('a'));
pub static KEY_B: Lazy<CaesarKey> = Lazy::new(|| CaesarKey::new('b'));

#[test]
fn key_try_new() {
  let valid_keys = LATIN_LOWERCASE.iter();
  let invalid_keys = RUSSIAN_LOWERCASE
    .iter()
    .chain(GREEK_LOWERCASE)
    .chain(DIGITS)
    .chain(SYMBOLS);

  valid_keys.for_each(|&c| assert!(CaesarKey::try_new(c, &CONTEXT).is_ok()));
  invalid_keys.for_each(|&c| assert!(CaesarKey::try_new(c, &CONTEXT).is_err()));
}

cipher_test!(CONTEXT, *QUICK_BROWN_FOX, KEY_A, latin_a_spaces);
cipher_test!(CONTEXT, *QUICKBROWNFOX, KEY_A, latin_a_no_spaces);
cipher_test!(CONTEXT, *QUICK_BROWN_FOX, KEY_B, latin_b_spaces);
cipher_test!(CONTEXT, *QUICKBROWNFOX, KEY_B, latin_b_no_spaces);
