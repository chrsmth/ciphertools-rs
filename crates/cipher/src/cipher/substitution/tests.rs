use super::*;
use crate::alphabet::Alphabet;
use crate::cipher_test;
use crate::tests::*;
use once_cell::sync::Lazy;

pub static CONTEXT: Lazy<Substitution> =
  Lazy::new(|| Substitution::new(Alphabet::latin()));

pub static KEY_A: Lazy<SubstitutionEncipherKey> = Lazy::new(|| {
  SubstitutionEncipherKey::new(
    "zebrascdfghijklmnopqtuvwxy".to_string(),
    &Alphabet::latin(),
  )
});

#[test]
fn key_try_new() {
  assert!(matches!(
    SubstitutionEncipherKey::try_new(
      "abcdefghijklmnopqrstuvwxy".to_string(),
      &CONTEXT
    ),
    Err(ParseError::LengthTooShort),
  ));

  assert!(matches!(
    SubstitutionEncipherKey::try_new(
      "abcdefghijklmnopqrstuvwxyza".to_string(),
      &CONTEXT
    ),
    Err(ParseError::LengthTooLong),
  ));

  assert!(matches!(
    SubstitutionEncipherKey::try_new(
      "abcdefghijklmnopqrs7uvwxyz".to_string(),
      &CONTEXT
    ),
    Err(ParseError::InvalidChar('7')),
  ));

  assert!(matches!(
    SubstitutionEncipherKey::try_new(
      "abcdefghijklmnopqrstuvwxya".to_string(),
      &CONTEXT
    ),
    Err(ParseError::DuplicateChar('a')),
  ));

  assert!(
    SubstitutionEncipherKey::try_new(
      "abcdefghijklmnopqrstuvwxyz".to_string(),
      &CONTEXT
    )
    .is_ok()
  );
}

cipher_test!(CONTEXT, *QUICK_BROWN_FOX, KEY_A, include_whitespace);
