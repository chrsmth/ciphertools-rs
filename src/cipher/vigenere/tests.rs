use super::*;
use crate::alphabet::Alphabet;
use crate::cipher_test;
use crate::tests::*;
use once_cell::sync::Lazy;

pub static CONTEXT_INCLUDE_WHITESPACE: Lazy<Vigenere> =
  Lazy::new(|| Vigenere::new(Alphabet::latin(), false));
pub static CONTEXT_SKIP_WHITESPACE: Lazy<Vigenere> =
  Lazy::new(|| Vigenere::new(Alphabet::latin(), true));

pub static KEY_A: Lazy<VigenereKey> =
  Lazy::new(|| VigenereKey::new("key".to_string()));

#[test]
fn key_try_new() {
  assert!(VigenereKey::try_new(
    "hello".to_string(),
    &CONTEXT_INCLUDE_WHITESPACE
  )
  .is_ok());
  assert!(matches!(
    VigenereKey::try_new("gr8ness".to_string(), &CONTEXT_INCLUDE_WHITESPACE),
    Err(ParseError::InvalidChar('8')),
  ));
}

cipher_test!(
  CONTEXT_INCLUDE_WHITESPACE,
  *QUICK_BROWN_FOX,
  KEY_A,
  include_whitespace
);

cipher_test!(
  CONTEXT_SKIP_WHITESPACE,
  *QUICK_BROWN_FOX,
  KEY_A,
  skip_whitespace
);
