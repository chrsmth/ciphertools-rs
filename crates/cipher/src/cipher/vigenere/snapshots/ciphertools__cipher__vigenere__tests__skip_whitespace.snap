---
source: src/cipher/vigenere/tests.rs
expression: ciphertext
---
dlc aygmo zbsux jmh nswtq yzcb xfo pyjc byk
